### ISSUE TITLE
Chord Type - Data Structure for new Difficulty UI

### ISSUE THREAD

**Issue by KCarlile** [KCarlile/flashchord#33]
## Background
I'm struggling with coming up with the right data structure for the chord types so that it works for specific keys, any key, and can map easily to the checkboxes on the UI. The current model doesn't really work. Each chord needs to relate to its function (major, minor, diminished, augmented, dominant, etc.) as well as its availability in the current session.

For reference, here's what the new UI will look like:
![image](https://user-images.githubusercontent.com/4503624/85498705-0eb94980-b595-11ea-88c0-5b2ce1d936ee.png)

## Acceptance Criteria
- The chord type data structure must allow for:
  - Filtering based on the current checkboxes selected in the UI
  - Filtering based on function (major, minor, diminished, augmented, etc.)
  - Usable in a specific key or in the Any key mode
- The UI must not inform the data structure (i.e. the data cannot come from the UI, only the settings can come from the UI) and the UI should reflect the data, not vice versa

## Proposed Approach
- Create data type templates for each chord type as an array
  - `major = ["", "maj7", "6", "sus2", "sus4"...];`
  - `minor = ["m", "m7", "m6", "m9"...];`
  - `diminished = ["dim", "dim7", "m7b5"...];`
  - `aug = ["+", "7alt"...];`
- On run, copy these to current versions and filter out the ones not selected in the checkboxes
  - `major_current = major;`
  - `delete major_current[major_current.indexOf('sus2')]; // repeat for each unchecked item`
  - `// repeat for minor_current, diminished_current, and aug_current`
- For specified keys, use the key>>chord quality lookup to select the correct array to choose from
- For any key, merge major_current, minor_current, diminished_current, and aug_current into one array

**Comment by KCarlile** [comment]
I think I can do this by building an array of all the checked chord types, then compare it against the data of available chords like this:
```
<script>
    $(function() {
        var $arr = ["a", "b", "c", "d", "e", "f", "g", "h", "i", "j", "k"];
        var $selections = ["b", "e", "k"];

        var $new = $arr.filter(function(chord) {
            return $selections.includes(chord);
        });

        $("#print").text($arr + ' ::::: ' + $selections + ' >>> ' + $new);
    });
</script>
```

For the extensions (`#5`, `b5`, `#9`, `b9`, etc.), I think I just need to check to see if the tone exists in the chord. That is, if it's trying to add b9 on a m9 chord, just don't add the extension or roll again for a new random extension.

**Comment by KCarlile** [comment]
Worded more betterer, on start...

1. Get all the checked chord types and put the values in an array
1. If Any key is selected, merge all 4 chord types into one array; else get the scale tone at random to determine the chord quality (major, minor, etc.)
1. Use the `array.filter()` function to compare the selected chord types against the possible chord types (in major, minor, diminished, augmented, and dominant arrays) to get only the valid chord types
1. If extensions, check to make sure we aren't trying to add a confusing extension, like `#9` on a m9 chord (maybe this is actually valid)

**Comment by KCarlile** [comment]
This was addressed in PR #41 and is testable at http://dev.flashchord.com/

**Comment by KCarlile** [comment]
Closed in release 1.5.
